docs/asciidoc/static/include/pluginbody.asciidoc

```diff
@@ -0,0 +1,1021 @@
+codec, and even an https://github.com/logstash-plugins/logstash-codec-avro/blob/master/lib/logstash/codecs/avro.rb#L38-L45[avro]
+codec.
+
+In most cases, your `encode` method should have an `@on_event.call()` statement.
+This call will output data per event in the described way.
+endif::encode_method[]
+
+// /////////////////////////////////////////////////////////////////////////////
+// If run_method is defined (should only be for input plugin page)
+// /////////////////////////////////////////////////////////////////////////////
+ifdef::run_method[]
+==== The `run` Method
+The `pass:attributes[{pluginname}]` input plugin has the following `run` Method:
+[source,ruby]
+[subs="attributes"]
+----------------------------------
+  def run(queue)
+    Stud.interval(@interval) do
+      event = LogStash::Event.new("message" => @message, "host" => @host)
+      decorate(event)
+      queue << event
+    end # loop
+  end # def run
+----------------------------------
+The `run` method is where a stream of data from an input becomes an event.
+
+The stream can be plain or generated as with the
+https://github.com/logstash-plugins/logstash-input-heartbeat/blob/master/lib/logstash/inputs/heartbeat.rb#L43-L61[heartbeat]
+input plugin.  In these cases, though no codec is used,
+https://github.com/logstash-plugins/logstash-input-heartbeat/blob/master/lib/logstash/inputs/heartbeat.rb#L17[a default codec]
+must be set in the code to avoid errors.
+
+Here's another example `run` method:
+[source,ruby]
+[subs="attributes"]
+----------------------------------
+  def run(queue)
+    while true
+      begin
+        # Based on some testing, there is no way to interrupt an IO.sysread nor
+        # IO.select call in JRuby.
+        data = $stdin.sysread(16384)
+        @codec.decode(data) do |event|
+          decorate(event)
+          event["host"] = @host if !event.include?("host")
+          queue << event
+        end
+      rescue IOError, EOFError, LogStash::ShutdownSignal
+        # stdin closed or a requested shutdown
+        break
+      end
+    end # while true
+    finished
+  end # def run
+----------------------------------
+In this example, we are sending the `data` to the codec defined in the
+configuration to `decode` the data stream and return an event.
+
+In both examples, the resulting `event` is passed to the `decorate` method:
+[source,ruby]
+[subs="attributes"]
+----------------------------------
+      decorate(event)
+----------------------------------
+This applies any tags you might have set in the input configuration block, e.g.
+`tags => ["tag1", "tag2"]`.
+
+Also in both examples, the `event`, after being "decorated," is appended to the
+queue:
+[source,ruby]
+[subs="attributes"]
+----------------------------------
+      queue << event
+----------------------------------
+
+This inserts the event into the pipeline.
+
+===== More Examples
```

Comment:
I've kind of lost track of the heading hierarchy at this point, but I think this is part of the anatomy section & should be boosted out. It's not part of the anatomy, it should be at the same level.